automates/model_analysis/graph_manipulation.py

```diff
@@ -0,0 +1,618 @@
+    return g_obs
+
+
+def unobserved_graph(g):
+    """
+    Constructs an unobserved graph
+    :param g: Input graph
+    :return: unobserved graph
+    """
+    g_copy = copy.deepcopy(g)
+    unobs_edges = g_copy.es.select(description="U")
+    u1 = len(unobs_edges)
+    if u1 > 0:
+        u = g_copy.es.select(description="U")
+        edges_to_remove = []
+        for edge in u:
+            edge_tuple = edge.tuple
+            if edge_tuple[1] > edge_tuple[0]:
+                edges_to_remove.append(edge.index)
+        g_copy.delete_edges(edges_to_remove)
+        e = g_copy.es.select(description="U")
+        e_len = len(e)
+        new_nodes = []
+        for i in range(e_len):
+            new_nodes.append(f"u_{{i + 1}}")
+        g_copy.add_vertices(new_nodes, attributes={"description": ["U"] * e_len})
+        edge_list = []
+
+        # We have now inserted new unobserved nodes into the graph
+        # We replace the unobserved bi-directed edges with new edges pointing away from the new unobserved nodes
+        # a is the index of the new unobserved node
+        # b and c are the two nodes that were previously connected by a bi-directed edge
+        for i in range(e_len):  # Loop through unobserved edges
+            a = g_copy.vs.select(name=new_nodes[i]).indices[0]
+            b = e[i].tuple[0]
+            edge_list.append((a, b))
+            c = e[i].tuple[1]
+            edge_list.append((a, c))
+        g_copy.add_edges(edge_list, attributes={"description": ["O"] * len(edge_list)})
+        obs_edges = g_copy.es.select(description_ne="U")
+        g_unobs = g_copy.subgraph_edges(obs_edges, delete_vertices=False)
+        return g_unobs
+    return g
+
+
+def ts(nodes, topo_order):  # topo must be a list of names
+    """
+    Orders nodes by their topological order
+    :param nodes: Nodes to be ordered
+    :param topo_order: Order to arrange nodes
+    :return: Ordered nodes (indices)
+    """
+    node_set = set(nodes)
+    return [n for n in topo_order if n in node_set]
+
+
+def ancestors(node, g, topo):
+    """
+    Finds all ancestors of a node and orders them
+    :param node: node (indicated by its index)
+    :param g: graph
+    :param topo: topological ordering
+    :return: Ancestors of node in topological ordering topo
+    """
+    an_list = g.neighborhood(node, order=g.vcount(), mode="in")
+    an_ind = list(set([a for ans in an_list for a in ans]))
+    an_names = to_names(an_ind, g)
+    an = ts(an_names, topo)
+    return an
+
+
+def ancestors_unsort(node, g):
+    """
+    Finds all ancestors of a node without the need for a topological ordering
+    :param node: set of nodes of which to find ancestors
+    :param g: graph
+    :return: Ancestors of nodes
+    """
+    an_list = g.neighborhood(node, order=g.vcount(), mode="in")
+    an_ind = list(set([a for ans in an_list for a in ans]))
+    an_names = to_names(an_ind, g)
+    return an_names
+
+
+def parents_unsort(node, g_obs):
```

Comment:
@aswinchester it looks like you can combine this function and `children_unsort()` into the `ancestors()` function by adding two more arguments: `order` and `mode`. You could name that function:
```python
find_related_nodes_of(node, g, order, mode, topo=None)
```
I think this would be the best schema for this function since it will require you to define a `order` and `mode` for each function call and that will help document the call as to whether we are finding parents, children, or ancestors. Does that sound good to you?

Reply:
I have created the function:
```
find_related_nodes_of(nodes, g, mode, order=1,  topo=None, exclude_orig=False):
```
and removed all of the old functions and updated their usages.